### PR TITLE
The first task is completed

### PR DESCRIPTION
**Pull request by IsaevMH**

### CODE REVIEW

**Comment by ZeroShadows** [comment]
pom.xml comments:
1. why it is already in master? I need to see changes in pom.xml in this pool request
2. lets move version of maven-compiler-plugin to properties section. Also source and target versions have to be moved into properties section.
3. lets be more specific into profile names. hw1-first-profile - > what should be tested into this profile?

**Comment by ZeroShadows** [comment]
Why did you decide to keep this src/main/java/com/epam/tc/hw1/MainApp.java into repository? From my point of view it does not apply to homework

**Comment by IsaevMH** [comment]
> 
> 
> pom.xml comments:
> 
>     1. why it is already in master? I need to see changes in pom.xml in this pool request
> 
>     2. lets move version of maven-compiler-plugin to properties section. Also source and target versions have to be moved into properties section.
> 
>     3. lets be more specific into profile names. hw1-first-profile - > what should be tested into this profile?

1. It's my mistake.
2. Issue is fixed
3. Issue is fixed

**Comment by IsaevMH** [comment]
> 
> 
> Why did you decide to keep this src/main/java/com/epam/tc/hw1/MainApp.java into repository? From my point of view it does not apply to homework

I guess it happens because I created the first project with the "Quickstart" archetype.
That issue is fixed

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+import com.epam.tat.module4.Calculator;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class DivisionCalcTest {
+
+    private Calculator calculator;
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:12]: lets create some sort of BaseTest and move Calculator initialisation there.

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    @BeforeClass(alwaysRun = true)
+    public void setUp() {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:15]: lets move test setup into BaseTest

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        calculator = new Calculator();
+    }
+
+    @Test(groups = {"MultAndDiv"}, dataProviderClass = DataProviderTest.class,
+            dataProvider = "correct data")
+    public void divisionTest(long firstDigit, long secondDigit) {
+        long resultOfDivision = calculator.div(firstDigit, secondDigit);
+        assertThat(resultOfDivision)
+                .isEqualTo(firstDigit / secondDigit);
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:24]: lets add expected result into data provider.
We have calculator under test, not JVM

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void afterClass() {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:28]: lets move test teardown into BaseTest

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package com.epam.tc.hw1;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+import com.epam.tat.module4.Calculator;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class MultiplicationCalcTest {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java:10]: extend this test class from BaseTest

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.epam.tc.hw1;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+import com.epam.tat.module4.Calculator;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class DivisionCalcTest {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:10]: extend this from BaseTest (you should create it first)

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    private Calculator calculator;
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java:12]: move Calculator initialisation into BaseTest

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+    @BeforeClass(alwaysRun = true)
+    public void setUp() {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java:15]: move test setup into BaseTest

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        calculator = new Calculator();
+    }
+
+    @Test(groups = {"MultAndDiv"}, dataProviderClass = DataProviderTest.class,
+            dataProvider = "correct data")
+    public void multiplicationTest(long firstDigit, long secondDigit) {
+        long resultOfMultiplication = calculator.mult(firstDigit, secondDigit);
+        assertThat(resultOfMultiplication)
+                .isEqualTo(firstDigit * secondDigit);
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java:24]: add expected result into data provider. 
Here we have calculator under test, not JVM

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void closeObject() {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java:28]: move test teardown into BaseTest

File: src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package com.epam.tc.hw1;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+import com.epam.tat.module4.Calculator;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.Test;
+
+public class SubtractionCalcTest {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java:10]: extend it from BaseTest

File: src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    private Calculator calculator;
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java:11]: move calculator initialisation to BaseTest

File: src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+    @BeforeTest(alwaysRun = true)
+    public void setUp() {
+        calculator = new Calculator();
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java:15]: move test initialisation into BaseTest

File: src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+            dataProvider = "correct data")
+    public void subtractionTest(long firstDigit, long secondDigit) {
+        long resultOfSubtraction = 0L;
+        resultOfSubtraction = calculator.sub(firstDigit, secondDigit);
+        assertThat(resultOfSubtraction)
+                .isEqualTo(firstDigit - secondDigit);
+    }
+
+    @AfterTest(alwaysRun = true)
+    public void closeObject() {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java:28]: move test teardown into BaseTest

File: src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+        calculator = new Calculator();
+    }
+
+    @Test(groups = {"SumAndSub"}, dataProviderClass = DataProviderTest.class,
+            dataProvider = "correct data")
+    public void subtractionTest(long firstDigit, long secondDigit) {
+        long resultOfSubtraction = 0L;
+        resultOfSubtraction = calculator.sub(firstDigit, secondDigit);
+        assertThat(resultOfSubtraction)
+                .isEqualTo(firstDigit - secondDigit);
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java:24]: add expected result into data provider
Here we have calculator under test, not JVM

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
+package com.epam.tc.hw1;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+import com.epam.tat.module4.Calculator;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class SummationCalcTest {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SummationCalcTest.java:10]: extend it from BaseTest

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    private Calculator calculator;
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SummationCalcTest.java:11]: move it to BaseTest

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+    @BeforeClass(alwaysRun = true)
+    public void setUp() {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SummationCalcTest.java:14]: move test initialisation into BaseTest

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+        calculator = new Calculator();
+    }
+
+    @Test(groups = {"SumAndSub"}, dataProviderClass = DataProviderTest.class,
+        dataProvider = "correct data")
+    public void summationTest(long firstDigit, long secondDigit) {
+        long resultOfSumming = calculator.sum(firstDigit, secondDigit);
+        assertThat(resultOfSumming)
+                .isEqualTo(firstDigit + secondDigit);
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SummationCalcTest.java:23]: add expected result into data provider.
Here we have calculator under test, not JVM

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void closeObject() {
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SummationCalcTest.java:27]: move test teardown to BaseTest

File: src/test/java/com/epam/tc/hw1/DataProviderTest.java [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+import org.testng.annotations.DataProvider;
+
+public class DataProviderTest {
+
+    @DataProvider(name = "correct data")
+    public static Object[][] correctData() {
+        return new Object[][] {
+                {6, 2},
+                {-4, 2},
+                {-8, -4}
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/DataProviderTest.java:12]: from my point of view, that's not enough for testing. lets add some extra test data.
or an extra data providers

**Review by ZeroShadows — COMMENTED** [review]
general comment:
from my point of view, data provider classes and test classes should be placed into different directories.
BaseTest can be placed into src/test/java/com/epam/tc/hw1

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java (hunk 19 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/SummationCalcTest.java:23]: How I can add the expected result?
Those data use for summation, division, multiplication and subtraction. After each operation, we expect a different result.
I need your help in that case.

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java (hunk 1 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:12]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java (hunk 2 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:15]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java (hunk 3 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:24]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java (hunk 4 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:28]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java (hunk 5 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java:10]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java (hunk 17 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/SummationCalcTest.java:11]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java (hunk 18 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/SummationCalcTest.java:14]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java (hunk 19 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/SummationCalcTest.java:23]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java (hunk 20 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/SummationCalcTest.java:27]: Issue is fixed

File: src/test/java/com/epam/tc/hw1/DataProviderTest.java (hunk 21 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/DataProviderTest.java:12]: Issue is fixed

**Comment by IsaevMH** [comment]
> 
> 
> general comment:
> from my point of view, data provider classes and test classes should be placed into different directories.
> BaseTest can be placed into src/test/java/com/epam/tc/hw1

Bugs were fixed

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java (hunk 6 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/DivisionCalcTest.java]: Issue is fixed

**Review by IsaevMH — COMMENTED** [review]
All bugs were fixed

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+import com.epam.tc.hw1.providers.DataProviderTest;
+import org.testng.annotations.Test;
+
+public class DivisionCalcTest extends BaseTest {
+
+    @Test(groups = {"MultAndDiv"}, dataProviderClass = DataProviderTest.class,
+            dataProvider = "division")
+    public void divisionTest(long firstDigit, long secondDigit, long expectedResult) {
+        long resultOfDivision = calculator.div(firstDigit, secondDigit);
+        assertThat(resultOfDivision)
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/DivisionCalcTest.java:14]: let's add 'as' with message to describe if assertion has failed

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+import com.epam.tc.hw1.providers.DataProviderTest;
+import org.testng.annotations.Test;
+
+public class MultiplicationCalcTest extends BaseTest {
+
+    @Test(groups = {"MultAndDiv"}, dataProviderClass = DataProviderTest.class,
+            dataProvider = "multiplication")
+    public void multiplicationTest(long firstDigit, long secondDigit, long expectedResult) {
+        long resultOfMultiplication = calculator.mult(firstDigit, secondDigit);
+        assertThat(resultOfMultiplication)
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java:14]: let's add 'as' with message to describe if assertion has failed

File: src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+import com.epam.tc.hw1.providers.DataProviderTest;
+import org.testng.annotations.Test;
+
+public class SubtractionCalcTest extends BaseTest {
+
+    @Test(groups = {"SumAndSub"}, dataProviderClass = DataProviderTest.class,
+            dataProvider = "subtraction")
+    public void subtractionTest(long firstDigit, long secondDigit, long expectedResult) {
+        long resultOfSubtraction = calculator.sub(firstDigit, secondDigit);
+        assertThat(resultOfSubtraction)
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java:14]: let's add 'as' with message to describe if assertion has failed

File: src/test/java/com/epam/tc/hw1/SummationCalcTest.java [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+import com.epam.tc.hw1.providers.DataProviderTest;
+import org.testng.annotations.Test;
+
+public class SummationCalcTest extends BaseTest {
+
+    @Test(groups = {"SumAndSub"}, dataProviderClass = DataProviderTest.class,
+        dataProvider = "summation")
+    public void summationTest(long firstDigit, long secondDigit, long expectedResult) {
+        long resultOfSumming = calculator.sum(firstDigit, secondDigit);
+        assertThat(resultOfSumming)
```
Comment (ZeroShadows) [inline src/test/java/com/epam/tc/hw1/SummationCalcTest.java:14]: let's add 'as' with message to describe if assertion has failed

Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/SummationCalcTest.java]: Issue was fixed

File: src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java (hunk 24 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/SubtractionCalcTest.java]: Issue was fixed

File: src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java (hunk 23 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/MultiplicationCalcTest.java]: Issue was fixed

File: src/test/java/com/epam/tc/hw1/DivisionCalcTest.java (hunk 22 above)
Comment (IsaevMH) [reply src/test/java/com/epam/tc/hw1/DivisionCalcTest.java]: Issue was fixed

**Pull request merged**